pac4j-saml-db/src/main/java/org/pac4j/saml/dbclient/DbLoadedSamlClients.java

```diff
@@ -0,0 +1,198 @@
+package org.pac4j.saml.dbclient;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.pac4j.core.client.Client;
+import org.pac4j.core.client.Clients;
+import org.pac4j.core.context.WebContext;
+import org.pac4j.core.exception.TechnicalException;
+import org.pac4j.core.util.CommonHelper;
+import org.pac4j.core.util.InitializableWebObject;
+import org.pac4j.saml.dbclient.dao.api.SamlClientDao;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * <p>
+ * An alternative to {@link Clients}. Unlike {@link Clients}, this class works with clients whose definition is stored in a database. It is
+ * limited to {@link DbLoadedSamlClient}s only, so clients of another type cannot be managed by this class.
+ * </p>
+ * 
+ * <p>
+ * This class is made to group multiple clients using a specific parameter to distinguish them, generally on one callback url.
+ * </p>
+ * 
+ * <p>
+ * The {@link #init(WebContext)} method is used to initialize the callback urls of the clients from the callback url of the clients group
+ * if empty and a specific parameter added to define the client targeted. It is implicitly called by the "finders" methods and doesn't
+ * need to be called explicitly.
+ * </p>
+ * 
+ * <p>
+ * The {@link #findClient(WebContext)}, {@link #findClient(WebContext, String)} or {@link #findClient(WebContext, Class)} methods must be
+ * called to find the right client according to the input context or type. The {@link #findAllClients(WebContext)} method returns all the
+ * clients.
+ * </p>
+ *
+ * TODO: We could have a common class (maybe abstract) that would contain common code of this class and {@link Clients}. Most code is
+ * duplicated. {@link Clients} cannot be extended because it is final.
+ * 
+ * @author jkacer
+ */
+public class DbLoadedSamlClients extends InitializableWebObject {

```

Comment:
Done, although not exactly as you wished. I removed all duplicate code but kept the new class. It extends the existing Clients class, so it has all its capabilities. It only overrides the internalInit() method + a few setters. So I believe the new class can be used anywhere where the old one is used now, bringing the option to load SAML clients in addition to defining clients statically.

If you really do not like this approach, then I would suggest to add something like "Dynamic Client Contributors" to the Clients class. It would have a list of contributors, where each would be allowed to add some clients to the list of managed clients. The addition would happen in internalInit(), similarly to the current implementation in DbLoadedSamlClients.